manuals/fshfsmount.1
.Dd July 30, 2026
.Dt FSHFSMOUNT 1
.Os
.Sh NAME
.Nm fshfsmount
.Nd mounts a Hierarchical File System (HFS) volume
.Sh SYNOPSIS
.Nm fshfsmount
.Op Fl o Ar offset
.Op Fl X Ar extended_options
.Op Fl hvV
.Ar source
.Ar mount_point
.Sh DESCRIPTION
.Nm fshfsmount
is a utility to mount a Hierarchical File System (HFS) volume
.Pp
.Nm fshfsmount
is part of the
.Nm libfshfs
package.
.Nm libfshfs
is a library to access the Hierarchical File System (HFS) format
.Pp
.Ar source
a Hierarchical File System (HFS) volume
.Ar mount_point
the directory to serve as mount point
.Pp
The options are as follows:
.Bl -tag -width Ds
.It Fl h
shows this help
.It Fl o Ar offset
specify the volume offset in bytes
.It Fl v
verbose output to stderr, while fshfsmount will remain running in the foreground
.It Fl V
print version
.It Fl X Ar extended_options
extended options to pass to sub system
.El
.Sh ENVIRONMENT
None
.Sh FILES
None
.Sh EXAMPLES
.Bd -literal
# fshfsmount /dev/sda1 /mnt/fshfs
fshfsmount 20260730
.sp
.Ed
.Sh DIAGNOSTICS
Errors, verbose and debug output are printed to stderr when verbose output \
\-v is enabled.
Verbose and debug output are only printed when enabled at compilation.
.Sh SEE ALSO
.Xr fshfsinfo 1
.Sh AUTHORS
.An Joachim Metz <joachim.metz@gmail.com>
.Sh BUGS
Please report bugs of any kind on the project issue tracker: \
https://github.com/libyal/libfshfs/issues
.Sh COPYRIGHT
Copyright (C) 2009-2026, Joachim Metz <joachim.metz@gmail.com>.
.sp
This is free software; see the source for copying conditions.
There is NO warranty; not even for MERCHANTABILITY or FITNESS FOR A \
PARTICULAR PURPOSE.


ChangeLog
To do:
* rename volume name to volume label ?

* check if get directory entries uses previous/next node

* add support for extended attribute (named fork) data stream with LIBFSHFS_ATTRIBUTE_RECORD_TYPE_EXTENTS
* for extended attributes by name look up use native B-tree structures

* info handle implement is_empty ?

* handle (hard) link count differently for indirect node files in /␀␀␀␀HFS+ Private Data/
* what about a symbolic link to HFS+ Private data directory?

* on file_entry_initialize lookup com.apple.decmpfs specifically instead of all attributes?
* catalog B-tree file, include name in branch key check
* file entry API: add extents functions

* add support to print information about special files (those without a catalog record) ?

* HFS+/HFSX
** add support for solaris and bsdos device numbers

* classic HFS
** fshfsinfo: add support retrieving non-directory file entries using fshfsinfo -E
*** files in classic HFS do not need to have a thread record
** API: add option to set file system encoding
** API: add means to determine if date and time values are in local time or UTC
** API: determine how to represent file mode
** add support for Mac OS codepages
** pass IO handle to directory, file and thread records to validate record types ?
** correct file record extents with extents start

Format support:
* implement support for classic HFS pre Mac OS 9.0
* implement support for HFS+/HFX Unicode 2.1 version
* implement support for folder and file information
* implement support for extended folder and file information

20160605
* see `git log' for more recent change log
* worked on format support

20160604
* worked on format support

20160603
* worked on format support

20160228
* worked on format support

20160227
* worked on format support

20160226
* worked on format support

20160220
* worked on format support

20160217
* 2016 update

20090212
* Initial version



fshfstools/fshfsmount.c
/*
 * Mounts a Hierarchical File System (HFS) volume.
 *
 * Copyright (C) 2009-2026, Joachim Metz <joachim.metz@gmail.com>
 *
 * Refer to AUTHORS for acknowledgements.
 *
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the GNU Lesser General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU Lesser General Public License
 * along with this program.  If not, see <https://www.gnu.org/licenses/>.
 */

#include <common.h>
#include <file_stream.h>
#include <memory.h>
#include <system_string.h>
#include <types.h>

#include <stdio.h>

#if defined( HAVE_FCNTL_H ) || defined( WINAPI )
#include <fcntl.h>
#endif

#if defined( HAVE_IO_H ) || defined( WINAPI )
#include <io.h>
#endif

#if defined( HAVE_STDLIB_H ) || defined( WINAPI )
#include <stdlib.h>
#endif

#if defined( HAVE_UNISTD_H )
#include <unistd.h>
#endif

#include "fshfstools_getopt.h"
#include "fshfstools_i18n.h"
#include "fshfstools_libcerror.h"
#include "fshfstools_libclocale.h"
#include "fshfstools_libcnotify.h"
#include "fshfstools_libfshfs.h"
#include "fshfstools_output.h"
#include "fshfstools_signal.h"
#include "fshfstools_unused.h"
#include "mount_dokan.h"
#include "mount_fuse.h"
#include "mount_handle.h"

mount_handle_t *fshfsmount_mount_handle = NULL;
int fshfsmount_abort                    = 0;

/* Signal handler for fshfsmount
 */
void fshfsmount_signal_handler(
      fshfstools_signal_t signal FSHFSTOOLS_ATTRIBUTE_UNUSED )
{
	libcerror_error_t *error = NULL;
	static char *function    = "fshfsmount_signal_handler";

	FSHFSTOOLS_UNREFERENCED_PARAMETER( signal )

	fshfsmount_abort = 1;

	if( fshfsmount_mount_handle != NULL )
	{
		if( mount_handle_signal_abort(
		     fshfsmount_mount_handle,
		     &error ) != 1 )
		{
			libcnotify_printf(
			 "%s: unable to signal mount handle to abort.\n",
			 function );

			libcnotify_print_error_backtrace(
			 error );
			libcerror_error_free(
			 &error );
		}
	}
	/* Force stdin to close otherwise any function reading it will remain blocked
	 */
#if defined( WINAPI ) && !defined( __CYGWIN__ )
	if( _close(
	     0 ) != 0 )
#else
	if( close(
	     0 ) != 0 )
#endif
	{
		libcnotify_printf(
		 "%s: unable to close stdin.\n",
		 function );
	}
}

/* The main program
 */
#if defined( HAVE_WIDE_SYSTEM_CHARACTER )
int wmain( int argc, wchar_t * const argv[] )
#else
int main( int argc, char * const argv[] )
#endif
{
	const char *description = \
		"Use fshfsmount to mount a Hierarchical File System (HFS) volume.";

	fshfstools_option_t options[ ] = {
		{ 'h', NULL, "shows this help" },
		{ 'o', "offset", "specify the volume offset in bytes" },
		{ 'v', NULL, "verbose output to stderr, while fshfsmount will remain running in the foreground" },
		{ 'V', NULL, "print version" },
#if defined( HAVE_LIBFUSE ) || defined( HAVE_LIBFUSE3 ) || defined( HAVE_LIBOSXFUSE )
		{ 'X', "extended_options", "extended options to pass to sub system" },
#endif
		{ 0, "volume", "a Hierarchical File System (HFS) volume" },
		{ 0, "mount_point", "the directory to serve as mount point" },
	};
	system_character_t options_string[ 32 ];

	libfshfs_error_t *error                     = NULL;
	system_character_t *option_offset           = NULL;
	system_character_t *source                  = NULL;
	char *program                               = "fshfsmount";
	system_integer_t option                     = 0;
	int number_of_options                       = (int) ( sizeof( options ) / sizeof( fshfstools_option_t ) );
	int verbose                                 = 0;

#if defined( HAVE_LIBFUSE ) || defined( HAVE_LIBFUSE3 ) || defined( HAVE_LIBOSXFUSE ) || defined( HAVE_LIBDOKAN )
	system_character_t *mount_point             = NULL;
	int result                                  = 0;
#endif

#if defined( HAVE_LIBFUSE ) || defined( HAVE_LIBFUSE3 ) || defined( HAVE_LIBOSXFUSE )
	struct fuse_operations fshfsmount_fuse_operations;

	system_character_t *option_extended_options = NULL;

#if defined( HAVE_LIBFUSE3 )
	/* Need to set this to 1 even if there no arguments, otherwise this causes
	 * fuse: empty argv passed to fuse_session_new()
	 */
	char *fuse_argv[ 2 ]                        = { program, NULL };
	struct fuse_args fshfsmount_fuse_arguments  = FUSE_ARGS_INIT(1, fuse_argv);
#else
	struct fuse_args fshfsmount_fuse_arguments  = FUSE_ARGS_INIT(0, NULL);
	struct fuse_chan *fshfsmount_fuse_channel   = NULL;
#endif
	struct fuse *fshfsmount_fuse_handle         = NULL;

#elif defined( HAVE_LIBDOKAN )
	DOKAN_OPERATIONS fshfsmount_dokan_operations;
	DOKAN_OPTIONS fshfsmount_dokan_options;
#endif

#if defined( __MINGW32__ ) && defined( HAVE_MINGW_BINMODE )
	_setmode( _fileno( stdout ), _O_BINARY );
	_setmode( _fileno( stderr ), _O_BINARY );
#endif

	libcnotify_stream_set(
	 stderr,
	 NULL );
	libcnotify_verbose_set(
	 1 );

	if( libclocale_initialize(
	     "fshfstools",
	     &error ) != 1 )
	{
		fprintf(
		 stderr,
		 "Unable to initialize locale values.\n" );

		goto on_error;
	}
	if( fshfstools_output_initialize(
	     _IONBF,
	     &error ) != 1 )
	{
		fprintf(
		 stderr,
		 "Unable to initialize output settings.\n" );

		goto on_error;
	}
	fshfstools_output_version_fprint(
	 stdout,
	 program );

	if( fshfstools_getopt_get_options_string(
	     options,
	     number_of_options,
	     options_string,
	     32 ) != 1 )
	{
		fprintf(
		 stderr,
		 "Unable to determine options string.\n" );

		goto on_error;
	}
	while( ( option = fshfstools_getopt(
	                   argc,
	                   argv,
	                   options_string ) ) != (system_integer_t) -1 )
	{
		switch( option )
		{
			case (system_integer_t) '?':
			default:
				fprintf(
				 stderr,
				 "Invalid argument: %" PRIs_SYSTEM "\n",
				 argv[ optind - 1 ] );

				fshfstools_getopt_usage_fprint(
				 stdout,
				 program,
				 description,
				 options,
				 number_of_options );

				return( EXIT_FAILURE );

			case (system_integer_t) 'h':
				fshfstools_getopt_usage_fprint(
				 stdout,
				 program,
				 description,
				 options,
				 number_of_options );

				return( EXIT_SUCCESS );

			case (system_integer_t) 'o':
				option_offset = optarg;

				break;

			case (system_integer_t) 'v':
				verbose = 1;

				break;

			case (system_integer_t) 'V':
				fshfstools_output_copyright_fprint(
				 stdout );

				return( EXIT_SUCCESS );

#if defined( HAVE_LIBFUSE ) || defined( HAVE_LIBFUSE3 ) || defined( HAVE_LIBOSXFUSE )
			case (system_integer_t) 'X':
				option_extended_options = optarg;

				break;
#endif
		}
	}
	if( optind == argc )
	{
		fprintf(
		 stderr,
		 "Missing source volume.\n" );

		fshfstools_getopt_usage_fprint(
		 stdout,
		 program,
		 description,
		 options,
		 number_of_options );

		return( EXIT_FAILURE );
	}
	source = argv[ optind++ ];

	if( optind == argc )
	{
		fprintf(
		 stderr,
		 "Missing mount point.\n" );

		fshfstools_getopt_usage_fprint(
		 stdout,
		 program,
		 description,
		 options,
		 number_of_options );

		return( EXIT_FAILURE );
	}
#if defined( HAVE_LIBFUSE ) || defined( HAVE_LIBFUSE3 ) || defined( HAVE_LIBOSXFUSE ) || defined( HAVE_LIBDOKAN )
	mount_point = argv[ optind ];
#endif
	libcnotify_verbose_set(
	 verbose );
	libfshfs_notify_set_stream(
	 stderr,
	 NULL );
	libfshfs_notify_set_verbose(
	 verbose );

	if( mount_handle_initialize(
	     &fshfsmount_mount_handle,
	     &error ) != 1 )
	{
		fprintf(
		 stderr,
		 "Unable to initialize mount handle.\n" );

		goto on_error;
	}
	if( option_offset != NULL )
	{
		if( mount_handle_set_offset(
		     fshfsmount_mount_handle,
		     option_offset,
		     &error ) != 1 )
		{
			fprintf(
			 stderr,
			 "Unable to set volume offset.\n" );

			goto on_error;
		}
	}
	if( mount_handle_open(
	     fshfsmount_mount_handle,
	     source,
	     &error ) != 1 )
	{
		fprintf(
		 stderr,
		 "Unable to open source volume\n" );

		goto on_error;
	}
#if defined( HAVE_LIBFUSE ) || defined( HAVE_LIBFUSE3 ) || defined( HAVE_LIBOSXFUSE )
	if( option_extended_options != NULL )
	{
#if defined( HAVE_LIBFUSE3 )
		// fuse_opt_add_arg: Assertion `!args->argv || args->allocated' failed.
		fshfsmount_fuse_arguments.argc = 0;
		fshfsmount_fuse_arguments.argv = NULL;
#endif
		/* This argument is required but ignored
		 */
		if( fuse_opt_add_arg(
		     &fshfsmount_fuse_arguments,
		     "" ) != 0 )
		{
			fprintf(
			 stderr,
			 "Unable add fuse arguments.\n" );

			goto on_error;
		}
		if( fuse_opt_add_arg(
		     &fshfsmount_fuse_arguments,
		     "-o" ) != 0 )
		{
			fprintf(
			 stderr,
			 "Unable add fuse arguments.\n" );

			goto on_error;
		}
		if( fuse_opt_add_arg(
		     &fshfsmount_fuse_arguments,
		     option_extended_options ) != 0 )
		{
			fprintf(
			 stderr,
			 "Unable add fuse arguments.\n" );

			goto on_error;
		}
	}
	if( memory_set(
	     &fshfsmount_fuse_operations,
	     0,
	     sizeof( struct fuse_operations ) ) == NULL )
	{
		fprintf(
		 stderr,
		 "Unable to clear fuse operations.\n" );

		goto on_error;
	}
	fshfsmount_fuse_operations.open       = &mount_fuse_open;
	fshfsmount_fuse_operations.read       = &mount_fuse_read;
	fshfsmount_fuse_operations.release    = &mount_fuse_release;
	fshfsmount_fuse_operations.getxattr   = &mount_fuse_getxattr;
	fshfsmount_fuse_operations.listxattr  = &mount_fuse_listxattr;
	fshfsmount_fuse_operations.opendir    = &mount_fuse_opendir;
	fshfsmount_fuse_operations.readdir    = &mount_fuse_readdir;
	fshfsmount_fuse_operations.releasedir = &mount_fuse_releasedir;
	fshfsmount_fuse_operations.getattr    = &mount_fuse_getattr;
	fshfsmount_fuse_operations.readlink   = &mount_fuse_readlink;
	fshfsmount_fuse_operations.destroy    = &mount_fuse_destroy;

#if defined( HAVE_LIBFUSE3 )
	fshfsmount_fuse_handle = fuse_new(
	                          &fshfsmount_fuse_arguments,
	                          &fshfsmount_fuse_operations,
	                          sizeof( struct fuse_operations ),
	                          fshfsmount_mount_handle );

	if( fshfsmount_fuse_handle == NULL )
	{
		fprintf(
		 stderr,
		 "Unable to create fuse handle.\n" );

		goto on_error;
	}
	result = fuse_mount(
	          fshfsmount_fuse_handle,
	          mount_point );

	if( result != 0 )
	{
		fprintf(
		 stderr,
		 "Unable to fuse mount file system.\n" );

		goto on_error;
	}
#else
	fshfsmount_fuse_channel = fuse_mount(
	                           mount_point,
	                           &fshfsmount_fuse_arguments );

	if( fshfsmount_fuse_channel == NULL )
	{
		fprintf(
		 stderr,
		 "Unable to create fuse channel.\n" );

		goto on_error;
	}
	fshfsmount_fuse_handle = fuse_new(
	                          fshfsmount_fuse_channel,
	                          &fshfsmount_fuse_arguments,
	                          &fshfsmount_fuse_operations,
	                          sizeof( struct fuse_operations ),
	                          fshfsmount_mount_handle );

	if( fshfsmount_fuse_handle == NULL )
	{
		fprintf(
		 stderr,
		 "Unable to create fuse handle.\n" );

		goto on_error;
	}
#endif /* defined( HAVE_LIBFUSE3 ) */

	if( verbose == 0 )
	{
		if( fuse_daemonize(
		     0 ) != 0 )
		{
			fprintf(
			 stderr,
			 "Unable to daemonize fuse.\n" );

			goto on_error;
		}
	}
	result = fuse_loop(
	          fshfsmount_fuse_handle );

	if( result != 0 )
	{
		fprintf(
		 stderr,
		 "Unable to run fuse loop.\n" );

		goto on_error;
	}
	fuse_unmount(
	 fshfsmount_fuse_handle );

	fuse_destroy(
	 fshfsmount_fuse_handle );

	fuse_opt_free_args(
	 &fshfsmount_fuse_arguments );

	return( EXIT_SUCCESS );

#elif defined( HAVE_LIBDOKAN )
	if( memory_set(
	     &fshfsmount_dokan_operations,
	     0,
	     sizeof( DOKAN_OPERATIONS ) ) == NULL )
	{
		fprintf(
		 stderr,
		 "Unable to clear dokan operations.\n" );

		goto on_error;
	}
	if( memory_set(
	     &fshfsmount_dokan_options,
	     0,
	     sizeof( DOKAN_OPTIONS ) ) == NULL )
	{
		fprintf(
		 stderr,
		 "Unable to clear dokan options.\n" );

		goto on_error;
	}
	fshfsmount_dokan_options.Version    = DOKAN_VERSION;
	fshfsmount_dokan_options.MountPoint = mount_point;

#if DOKAN_MINIMUM_COMPATIBLE_VERSION >= 200
	fshfsmount_dokan_options.SingleThread = TRUE;
#else
	fshfsmount_dokan_options.ThreadCount  = 0;
#endif
	if( verbose != 0 )
	{
		fshfsmount_dokan_options.Options |= DOKAN_OPTION_STDERR;
#if defined( HAVE_DEBUG_OUTPUT )
		fshfsmount_dokan_options.Options |= DOKAN_OPTION_DEBUG;
#endif
	}
/* This will only affect the drive properties
	fshfsmount_dokan_options.Options |= DOKAN_OPTION_REMOVABLE;
*/

#if ( DOKAN_VERSION >= 600 ) && ( DOKAN_VERSION < 800 )
	fshfsmount_dokan_options.Options |= DOKAN_OPTION_KEEP_ALIVE;

	fshfsmount_dokan_operations.CreateFile           = &mount_dokan_CreateFile;
	fshfsmount_dokan_operations.OpenDirectory        = &mount_dokan_OpenDirectory;
	fshfsmount_dokan_operations.CreateDirectory      = NULL;
	fshfsmount_dokan_operations.Cleanup              = NULL;
	fshfsmount_dokan_operations.CloseFile            = &mount_dokan_CloseFile;
	fshfsmount_dokan_operations.ReadFile             = &mount_dokan_ReadFile;
	fshfsmount_dokan_operations.WriteFile            = NULL;
	fshfsmount_dokan_operations.FlushFileBuffers     = NULL;
	fshfsmount_dokan_operations.GetFileInformation   = &mount_dokan_GetFileInformation;
	fshfsmount_dokan_operations.FindFiles            = &mount_dokan_FindFiles;
	fshfsmount_dokan_operations.FindFilesWithPattern = NULL;
	fshfsmount_dokan_operations.SetFileAttributes    = NULL;
	fshfsmount_dokan_operations.SetFileTime          = NULL;
	fshfsmount_dokan_operations.DeleteFile           = NULL;
	fshfsmount_dokan_operations.DeleteDirectory      = NULL;
	fshfsmount_dokan_operations.MoveFile             = NULL;
	fshfsmount_dokan_operations.SetEndOfFile         = NULL;
	fshfsmount_dokan_operations.SetAllocationSize    = NULL;
	fshfsmount_dokan_operations.LockFile             = NULL;
	fshfsmount_dokan_operations.UnlockFile           = NULL;
	fshfsmount_dokan_operations.GetFileSecurity      = NULL;
	fshfsmount_dokan_operations.SetFileSecurity      = NULL;
	fshfsmount_dokan_operations.GetDiskFreeSpace     = NULL;
	fshfsmount_dokan_operations.GetVolumeInformation = &mount_dokan_GetVolumeInformation;
	fshfsmount_dokan_operations.Unmount              = &mount_dokan_Unmount;

#else
	fshfsmount_dokan_operations.ZwCreateFile         = &mount_dokan_ZwCreateFile;
	fshfsmount_dokan_operations.Cleanup              = NULL;
	fshfsmount_dokan_operations.CloseFile            = &mount_dokan_CloseFile;
	fshfsmount_dokan_operations.ReadFile             = &mount_dokan_ReadFile;
	fshfsmount_dokan_operations.WriteFile            = NULL;
	fshfsmount_dokan_operations.FlushFileBuffers     = NULL;
	fshfsmount_dokan_operations.GetFileInformation   = &mount_dokan_GetFileInformation;
	fshfsmount_dokan_operations.FindFiles            = &mount_dokan_FindFiles;
	fshfsmount_dokan_operations.FindFilesWithPattern = NULL;
	fshfsmount_dokan_operations.SetFileAttributes    = NULL;
	fshfsmount_dokan_operations.SetFileTime          = NULL;
	fshfsmount_dokan_operations.DeleteFile           = NULL;
	fshfsmount_dokan_operations.DeleteDirectory      = NULL;
	fshfsmount_dokan_operations.MoveFile             = NULL;
	fshfsmount_dokan_operations.SetEndOfFile         = NULL;
	fshfsmount_dokan_operations.SetAllocationSize    = NULL;
	fshfsmount_dokan_operations.LockFile             = NULL;
	fshfsmount_dokan_operations.UnlockFile           = NULL;
	fshfsmount_dokan_operations.GetFileSecurity      = NULL;
	fshfsmount_dokan_operations.SetFileSecurity      = NULL;
	fshfsmount_dokan_operations.GetDiskFreeSpace     = NULL;
	fshfsmount_dokan_operations.GetVolumeInformation = &mount_dokan_GetVolumeInformation;
	fshfsmount_dokan_operations.Unmounted            = NULL;
	fshfsmount_dokan_operations.FindStreams          = NULL;
	fshfsmount_dokan_operations.Mounted              = NULL;

#endif /* ( DOKAN_VERSION >= 600 ) && ( DOKAN_VERSION < 800 ) */

#if DOKAN_MINIMUM_COMPATIBLE_VERSION >= 200
	DokanInit();
#endif
	result = DokanMain(
	          &fshfsmount_dokan_options,
	          &fshfsmount_dokan_operations );

#if DOKAN_MINIMUM_COMPATIBLE_VERSION >= 200
	DokanShutdown();
#endif
	switch( result )
	{
		case DOKAN_SUCCESS:
			break;

		case DOKAN_ERROR:
			fprintf(
			 stderr,
			 "Unable to run dokan main: generic error\n" );
			break;

		case DOKAN_DRIVE_LETTER_ERROR:
			fprintf(
			 stderr,
			 "Unable to run dokan main: bad drive letter\n" );
			break;

		case DOKAN_DRIVER_INSTALL_ERROR:
			fprintf(
			 stderr,
			 "Unable to run dokan main: unable to load driver\n" );
			break;

		case DOKAN_START_ERROR:
			fprintf(
			 stderr,
			 "Unable to run dokan main: driver error\n" );
			break;

		case DOKAN_MOUNT_ERROR:
			fprintf(
			 stderr,
			 "Unable to run dokan main: unable to assign drive letter\n" );
			break;

		case DOKAN_MOUNT_POINT_ERROR:
			fprintf(
			 stderr,
			 "Unable to run dokan main: mount point error\n" );
			break;

		default:
			fprintf(
			 stderr,
			 "Unable to run dokan main: unknown error: %d\n",
			 result );
			break;
	}
	return( EXIT_SUCCESS );

#else
	fprintf(
	 stderr,
	 "No sub system to mount Hierarchical File System (HFS) format.\n" );

	return( EXIT_FAILURE );

#endif /* defined( HAVE_LIBFUSE ) || defined( HAVE_LIBFUSE3 ) || defined( HAVE_LIBOSXFUSE ) */

on_error:
	if( error != NULL )
	{
		libcnotify_print_error_backtrace(
		 error );
		libcerror_error_free(
		 &error );
	}
#if defined( HAVE_LIBFUSE ) || defined( HAVE_LIBFUSE3 ) || defined( HAVE_LIBOSXFUSE )
	if( fshfsmount_fuse_handle != NULL )
	{
		fuse_destroy(
		 fshfsmount_fuse_handle );
	}
	fuse_opt_free_args(
	 &fshfsmount_fuse_arguments );
#endif
	if( fshfsmount_mount_handle != NULL )
	{
		mount_handle_free(
		 &fshfsmount_mount_handle,
		 NULL );
	}
	return( EXIT_FAILURE );
}

